### PR TITLE
Initial build and release workflows with Github Actions

### PR DESCRIPTION
**Pull request by langston-barrett**
These are some relatively minimal build and release workflows for SemTK. Some things to note:

 1. There's some duplicated work between the build and release workflows (they both build SemTK). Since releases are relatively infrequent, I don't think this is a huge issue.
 2. The tests are skipped. I tried enabling them, but they failed. I think it's fine to start with a build and try to debug the tests if SemTK developers find it worthwhile.

You can see examples of these workflows in action on my fork (after I pushed a fake `v0.0.1` tag):
- https://github.com/langston-barrett/semtk/releases
- https://github.com/langston-barrett/semtk/tags
- https://github.com/langston-barrett/semtk/actions

### CODE REVIEW

**Comment by langston-barrett** [comment]
I can't assign reviewers on this repo, but cc @cuddihyge, @tuxji, and @nikita-visnevski for review.

**Review by tuxji — COMMENTED** [review]
Some thoughts.  

First, I noticed a warning in the CI run.

```
Warning: Unexpected input(s) 'artifacts', valid inputs are ['tag_name', 'release_name', 'body', 'body_path', 'draft', 'prerelease', 'commitish', 'owner', 'repo']
Run actions/create-release@v1
  with:
    artifacts: distribution/target/semtk-opensource-refs/tags/v0.0.1-bin.tar.gz
    tag_name: refs/tags/v0.0.1
    release_name: Release refs/tags/v0.0.1
    draft: true
    prerelease: false
```

You probably want two actions: create-release and update-release-asset.

```
    - name: Upload README.md to GitHub release assets
      uses: actions/upload-release-asset@v1
      env:
        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
      with:
        upload_url: ${{ github.event.release.upload_url }}
        asset_path: docs/README.md
        asset_name: README.md
        asset_content_type: text/markdown; charset=UTF-8
```

Second, I'm not very enthusiastic about the GitHub actions calling bash scripts.  You can write multi-line scripts directly in the CI workflow files, and I don't think the commands are valuable enough to warrant putting them in their own files.   It's also much easier to understand what a CI workflow does if all the commands are visible in the CI workflow file instead of having to go look at another file.  Here's an example how to write a multi-line run step and CI will call bash with some variant of -euo pipefail so you don't have to do that either:

```
    - name: Assemble extern folder
      run: |
        mv tools/verdict-back-ends/STEM extern
        mkdir extern/examples
        mv models/CASE_Toy_Models extern/examples
        mv models/DeliveryDrone_CASE_v* extern/examples
        chmod a+x extern/{mac,nix}/{aadl2iml,kind2,soteria_pp}
        cp -a tools/verdict-back-ends/verdict-bundle/verdict-bundle-app/target/verdict-bundle-app-*-capsule.jar extern
        zip -r extern.zip extern
        mkdir tools/verdict-back-ends/{aadl2iml,soteria_pp}/bin
        cp -a extern/nix/aadl2iml tools/verdict-back-ends/aadl2iml/bin
        cp -a extern/nix/soteria_pp tools/verdict-back-ends/soteria_pp/bin
```

**Comment by langston-barrett** [comment]
> First, I noticed a warning in the CI run.

Nice catch - I'll try and fix this!

> Second, I'm not very enthusiastic about the GitHub actions calling bash scripts. 

I'm glad you brought this up. I have my own thoughts on this trade-off, I'd be interested to hear yours as well. 

**Advantages to putting commands in their own file**:

 - If the scripts are set up reasonably, you can more easily replicate the steps of the CI build locally by just running the script and seeing if it works
 - Editors can do things like highlight syntax and show syntax errors
 - Static analysis tools like `shellcheck` can help catch mistakes
 - The execution environment is more obvious (the shebang line mentions `bash`, and the options like `-eu` are set explicitly)
 - Quoting and variable references are less ambiguous. Admittedly, this isn't a huge problem with YAML's `|` blocks, but for instance Github Actions lets you put syntax like `${{variable}}` in some parts of your file, which looks a little too much like Bash's `${variable}` syntax, and could lead to confusion. I don't actually know if Github Actions would substitute a pattern like that in a `|` block, but I don't want to find out by making a confusing mistake.

**Advantages to embedding commands in the workflow file**:

 - Like you said, it's easier to tell what the whole workflow is doing, because all the steps are in one place
 - Also like you said, you don't have to repeat things like `set -euo pipefail` because they are implicitly set by Github Actions

I'd be happy to learn of more advantages of the second approach, or to hear that I'm weighing these pros and cons wrong and that the advantages to the YAML-embedding approach are more significant than the advantages of the separate script approach.

**Comment by tuxji** [comment]
I agree with what you said elsewhere that installing a real build system would reduce the CI workflow to just calling build steps anyway.  I think what I'm trying to say is that if there's no build system and you have to use GitHub Actions as a poor man's build system, then please dive into GitHub Actions with full commitment.  Try to write everything in there and take full advantage of any feature that might help.  Learn as much as possible about the syntax for workflow steps, search for available GitHub Marketplace actions, and even consider writing an action if it would simplify several steps.  The `${{variable}}` substitutions that GitHub Actions makes actually can be useful (they work in `run |` blocks too) and GitHub also has some [workflow commands](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions) with weird syntax (usually echo statements) that can do useful things.

However, the rest of your advantages make sense as well (testing CI steps locally, static analysis tools, etc.).

**Comment by langston-barrett** [comment]
I've inlined the script in CI. Since it's just two lines, I think the trade-offs are more on that side.

> I think what I'm trying to say is that if there's no build system and you have to use GitHub Actions as a poor man's build system, then please dive into GitHub Actions with full commitment. 

I hear you! RACK is the first project that I'm using Github Actions on, I'm eager to learn more about it.

**Comment by langston-barrett** [comment]
I fixed the warning and properly separated the build and upload steps. Recent builds can be seen here (the failures are from me using tags that didn't correspond to the tarball that actually gets output by the build process): https://github.com/langston-barrett/semtk/actions

**Comment by langston-barrett** [comment]
Whoops, I just noticed that the Release build didn't occur on my lastest push because the tag lacked a `v` prefix. Working on that now...

**Comment by langston-barrett** [comment]
> Whoops, I just noticed that the Release build didn't occur on my lastest push because the tag lacked a v prefix. Working on that now...

Okay, this is fixed now and this PR is ready for another round of review.

File: .github/workflows/build.yaml
Comment (tuxji) [inline .github/workflows/build.yaml:2]: Do we want this CI to run on a push to any branch, or only a push to the main branch?  Either is fine with me.  Paul, you can decide.

File: .github/workflows/build.yaml
Comment (tuxji) [inline .github/workflows/build.yaml:7]: It's a bummer that ubuntu-latest is still ubuntu-18.04, not ubuntu-20.04.  It doesn't matter here, but it matters over in RACK since the images are based on ubuntu-20.04 so I use ubuntu-20.04 in the CI workflow there too.

File: .github/workflows/build.yaml
Comment (tuxji) [inline .github/workflows/build.yaml:14]: The SemTK poms don't use environment variables, so there's no need to source .env here.

File: .github/workflows/release.yaml
Comment (tuxji) [inline .github/workflows/release.yaml:19]: Another thought about this build step.  We aren't caching ~/.m2 so Maven has to download a bunch of jars every time we run this build step.  If there wasn't a --quiet in the command, we would see the "Downloading" lines in the CI output and realize that we need a cache.  I'd remove the --quiet (the Maven output might be useful to look at sometime) and add a cache step before the build step like:

```
    - name: Cache Maven repository
      uses: actions/cache@v2
      with:
        path: ~/.m2
        key: maven-${{ hashFiles('**/pom.xml') }}
```

File: .github/workflows/release.yaml
Comment (tuxji) [inline .github/workflows/release.yaml:45]: Question for Langston - I looked at your repo fork's releases page, but none of your recent releases have the distribution tarball attached to them.  Did you remove the tarball manually after testing that these steps worked?

Question for Paul - do we want to create a release on the Releases page and attach the tarball there?  Or do we want to deploy to Sonatype so that snapshots will be available there and releases will be sync'ed to Maven central too?

File: .github/workflows/release.yaml
Comment (tuxji) [inline .github/workflows/release.yaml:5]: Question for Paul - how do you want to create a release?  You can make a tag and push it to git which will kick off this release workflow and release a semtk-opensource-{{tag}}-bin.tar.gz (everything else will be thrown away by the way - I ask you further below if you'd rather deploy everything to Sonatype).  Alternatively, you can go to the Releases page, click the `Draft a new release` button, choose whichever version/tag you want, fill out the rest of the form, and then submitting the form will kick off this workflow if we change its beginning from `on / push / tags` to 

```
on:
  release:
    types: [ published ]
```

**Review by tuxji — CHANGES_REQUESTED** [review]
Some questions and suggestions for improvements.

File: .github/workflows/build.yaml
Comment (langston-barrett) [reply .github/workflows/build.yaml:2]: cc @cuddihyge

Comment (langston-barrett) [reply .github/workflows/build.yaml:2]: I would recommend that we do on every push, so that developers can test how their code will do in CI before opening a merge request and sending a notification to all the people who "Watch" (in Github terms) this repo. Plus, it doesn't cost us anything :smile:

File: .github/workflows/build.yaml
Comment (langston-barrett) [reply .github/workflows/build.yaml:14]: Okay, thanks for the pointer! I'll try removing this.

File: .github/workflows/build.yaml
Comment (langston-barrett) [reply .github/workflows/build.yaml:7]: Sounds like no action needed here for now, then?

File: .github/workflows/release.yaml
Comment (langston-barrett) [reply .github/workflows/release.yaml:19]: > I'd remove the --quiet (the Maven output might be useful to look at sometime) and add a cache step before the build step like:

Will do. 

> I'd remove the --quiet (the Maven output might be useful to look at sometime)

I go back and forth on verbosity in CI logs: Often it's unhelpful and results in a lot of scrolling and grepping when a problem arises. On the other hand, as you say, it can be ocassionally useful. And Github Actions at least separates the logs between different steps, which is helpful.

File: .github/workflows/build.yaml
Comment (tuxji) [reply .github/workflows/build.yaml:7]: Right, I'm leaving ubuntu-latest vs. ubuntu-20.04 up to your choice (or Paul's).  I don't know why GitHub is taking so long to make ubuntu-latest the same as ubuntu-20.04 - it's been 6 months since 20.04 LTS came out.

File: .github/workflows/release.yaml
Comment (langston-barrett) [reply .github/workflows/release.yaml:45]: The very latest one has a distribution tarball attached:

https://github.com/langston-barrett/semtk/releases/tag/untagged-826ce881b5960e036f88

It's only the latest one because that's the only tag which generates an `asset_path` in this step that actually matches a build product and successfully uploads. Does that make sense?

Comment (tuxji) [reply .github/workflows/release.yaml:45]: Oh, your release is still in draft mode and hasn't been published yet.  I get a 404 when I click on your release link and I don't see your latest release in <https://github.com/langston-barrett/semtk/releases/>, only 2.2.2-SNAPSHOT (7cfc1b0).

Comment (langston-barrett) [reply .github/workflows/release.yaml:45]: Ah, I didn't realize that draft releases weren't publicly visible! This link should work: https://github.com/langston-barrett/semtk/releases/tag/2.2.2-SNAPSHOT

> Question for Paul - do we want to create a release on the Releases page and attach the tarball there? Or do we want to deploy to Sonatype so that snapshots will be available there and releases will be sync'ed to Maven central too?

My two cents: Why not both? They both have advantages for users.

- Github releases are easily accessible from the repo via the web interface, and are easy to set up with Actions. For users who are also using Github for other purposes in their CI, it's one fewer external service that has to be up for their pipelines/builds to work.
- Sonatype/Maven will be more familiar to a different audience (Java developers). For consumers of the SemTK Java APIs, this could result in one fewer external service that has to be up for their builds/pipelines to work.

The big disadvantage seems to be maintenance: Taking the time to _publish_ the Github releases, and repairing the upload steps if they ever become stale.

**Comment by langston-barrett** [comment]
Review comments addressed, just waiting on feedback from @cuddihyge.

**Comment by cuddihyge** [comment]
As @tuxji and I have discussed, we definitely want to auto deploy the semtk snapshot jar to sonatype.
As I don't yet understand these CI processes, I'd prefer that @tuxji or someone who understands them better merge the pull request.  I'd like a tutorial on these CI processes when possible.

**Comment by tuxji** [comment]
OK, then Langston, you can merge the pull request and we can do another PR which adds the ability to autodeploy to Sonatype later.  We also can refine exactly how one kicks off an autodeploy or release build later.  The second PR also will need to handle both snapshots (2.2.3-SNAPSHOT) and final releases (2.2.3) - only the latter kind will be copied to Maven central, so we will need to make final releases from time to time.

**Comment by langston-barrett** [comment]
> OK, then Langston, you can merge the pull request and we can do another PR which adds the ability to autodeploy to Sonatype later. 

Sounds good to me. I don't have push access to this repo so I'll need you or @cuddihyge to actually hit "merge". I've made #13 to track progress on Sonatype uploads.

> I'd like a tutorial on these CI processes when possible. 

Good idea. Let's do this after #13 is fixed, so we can go over the full picture.

**Pull request merged**
